### PR TITLE
Functional test performance improvement

### PR DESCRIPTION
**Pull request by fedys**
| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | staging
| Bug fix?                               | no
| New feature?                           | no
| Deprecations?                          | no
| BC breaks?                             | no
| Automated tests included?              | yes
| Related user documentation PR URL      |
| Related developer documentation PR URL |
| Issue(s) addressed                     | 

#### Description:
This PR just **improve speed/memory consumption** of functional tests.
- Before optimization: Time: 15.63 minutes, Memory: 1.14 GB
- After optimization: Time: 6.97 minutes, Memory: 709.00 MB

#### Steps to test this PR:
1. All automated tests must pass

#### The following points should be considered when writing functional tests:
* The biggest bottlenecks
    * Recreating whole DB before every test - solved via transaction rollback
    * Loading unnecessary fixtures (plus emptying data before, MauticWebTestCase loaded all fixtures including plugin fixtures in setUp() method)
    * Booting kernel multiple times within single test
    * Debug mode enabled
* Every test is responsible for a cleanup (restoring DB state)
    * Easiest way is to extend from MauticMysqlTestCase which uses a transaction rollback
* Sometimes transaction rollback cannot be used
    * Alter DB schema queries
    * Full-text queries don't see uncommitted changes
    * Need for booting kernel multiple times during single test
* Try avoiding loading fixtures
    * Very slow
    * Can be inflexible in some cases
    * More difficult to read than data creation done inside a given test
    * If you do load fixtures try appending them as removal is expensive
* Try avoiding data providers within functional tests
    * Booting kernel is expensive
* Good practice
    * Make sure your functional test cleans up DB properly
* Bad practices
    * relying on fixed ID values like 1,2,3
* Tips:
    * Set transaction isolation level to "Read uncomitted" in your DB client so you can see DB changes when debugging tests

### CODE REVIEW

**Review by escopecz — DISMISSED** [review]
Amazing results! 👍 

Imagine how much carbon this optimization will save 🍀

**Comment by RCheesley** [comment]
This is incredible, thank you @fedys for your work on this!

**Review by dennisameling — COMMENTED** [review]
Thanks SO much for this PR @fedys! Would you mind rebasing the PR so we can get it merged for the 3.2 release? Thanks in advance! 🚀

**Review by dennisameling — APPROVED** [review]
This looks fantastic. I especially like what you did with the database transactions + rollbacks, as well as the database fixtures. I'm planning to write a "getting started with writing tests for Mautic" guide and can take a lot of inspiration from this PR.

Thanks!! 🚀

File: app/bundles/CoreBundle/Test/MauticMysqlTestCase.php [hunk 1]
```diff
@@ -10,44 +12,105 @@
...
+
+    /**
+     * @throws Exception
      */
     protected function setUp()
     {
         parent::setUp();
 
-        $this->sqlDumpFile = $this->container->getParameter('kernel.cache_dir').'/fresh_db.sql';
+        if (!self::$databasePrepared) {
```
Comment (dennisameling) [inline app/bundles/CoreBundle/Test/MauticMysqlTestCase.php:35]: MASSIVE time-saver here. Thanks!

**Comment by fedys** [comment]
@dennisameling you are welcome! I thank @escopecz for finding [this repo](https://github.com/hautelook/AliceBundle#database-testing) that was my inspiration.

**Pull request merged**